caom2-tap/src/main/java/ca/nrc/cadc/caom2ops/AdCutoutGenerator.java

```diff
@@ -62,13 +63,51 @@ public URL toURL(URI uri, List<String> cutouts) {
         }
     }
 
+    static String removeCompressionExtension(String uriFilename) {
+        String filename = uriFilename;
+        int i = uriFilename.lastIndexOf('.');
+        if (i != -1 && i < uriFilename.length() - 1) {
+            String ext = uriFilename.substring(i + 1, uriFilename.length());
+            if (ext.equalsIgnoreCase("cf") || ext.equalsIgnoreCase("z")
+                || ext.equalsIgnoreCase("gz") || ext.equalsIgnoreCase("fz")) {
+                filename = uriFilename.substring(0, i);
+            }
+        }
+
+        return filename;
+    }
+
+    static String generateFilename(URI uri, String label) {
```

Comment:
This method needs to validate that the label is safe to add to a filename. I think CaomValidator.assertValidPathComponent does the checks that you would want here... it throws an IllegalArgumentException so you'll have to see how that shows up in the response. It probably needs to be turned into a UsageFault to get the required error message form

Reply:
Added validation.